### PR TITLE
Add --backend and --use-v2 support to detection refs

### PR DESCRIPTION
**Pull request by NicolasHug**

### CODE REVIEW

File: references/classification/presets.py [hunk 1]
```diff
@@ -30,42 +33,42 @@ def __init__(
         backend="pil",
         use_v2=False,
     ):
-        module = get_module(use_v2)
+        T = get_module(use_v2)
```
Comment (NicolasHug) [inline references/classification/presets.py:36]: I just did `s/module/T/` in the file to make it consistent with the detection one

File: references/detection/transforms.py [hunk 2]
```diff
@@ -293,11 +293,13 @@ def __init__(
         target_size: Tuple[int, int],
         scale_range: Tuple[float, float] = (0.1, 2.0),
         interpolation: InterpolationMode = InterpolationMode.BILINEAR,
+        antialias=True,
```
Comment (NicolasHug) [inline references/detection/transforms.py:296]: Had to add antialias support because it'd be False otherwise by default for tensors. There's no BC requirements so we could just hard-code `antialias=True` below in the calls to `resize()` instead of adding a parameter here, but it doesn't change much. LMK what you prefer.

File: references/detection/coco_utils.py [hunk 3]
```diff
@@ -220,25 +219,29 @@ def __getitem__(self, idx):
...
+        dataset = torchvision.datasets.CocoDetection(img_folder, ann_file, transforms=transforms)
+        # TODO: need to update target_keys to handle masks for segmentation!
+        dataset = wrap_dataset_for_transforms_v2(dataset, target_keys={"boxes", "labels", "image_id"})
+    else:
+        t = [ConvertCocoPolysToMask()]
+        if transforms is not None:
+            t.append(transforms)
+        transforms = T.Compose(t)
+
+        dataset = CocoDetection(img_folder, ann_file, transforms=transforms)
```
Comment (NicolasHug) [inline references/detection/coco_utils.py:244]: I wonder if we could get rid of this custom `CocoDetection` dataset here. Ideally we would always call `wrap_dataset_for_transforms_v2` and just "unwrap" the datapoints classes into pure-tensors etc...? But we can't use it without silencing the V2 warning first :/

Not sure what to do to clean that up.

File: references/detection/coco_utils.py [hunk 4]
```diff
@@ -126,10 +126,6 @@ def _has_valid_annotation(anno):
             return True
         return False
 
-    if not isinstance(dataset, torchvision.datasets.CocoDetection):
```
Comment (NicolasHug) [inline references/detection/coco_utils.py:129]: Instead of removing this (seemingly useless check) I could just add the same workaround as elsewhere i.e. add

```py
of isinstance(
        getattr(dataset, "_dataset", None), torchvision.datasets.CocoDetection
    ):
```

File: references/detection/presets.py [hunk 5]
```diff
@@ -1,73 +1,107 @@
...
+        T, datapoints = get_modules(use_v2)
+
+        transforms = []
+        backend = backend.lower()
+        if backend == "datapoint":
+            transforms.append(T.ToImageTensor())
+        elif backend == "tensor":
+            transforms.append(T.PILToTensor())
+        elif backend != "pil":
+            raise ValueError(f"backend can be 'datapoint', 'tensor' or 'pil', but got {backend}")
```
Comment (NicolasHug) [inline references/detection/presets.py:38]: I might need to explicitly convert to PIL, I don't remember (it depends on the wrapper). I'll check again later after the first round of reviews

File: references/detection/coco_utils.py [hunk 6]
```diff
@@ -196,12 +192,15 @@ def convert_to_coco_api(ds):
 
 
 def get_coco_api_from_dataset(dataset):
+    # FIXME: This is... awful?
```
Comment (pmeier) [inline references/detection/coco_utils.py:195]: Yeah. Happy for you to address it here, but not required.

File: references/detection/engine.py [hunk 7]
```diff
@@ -26,7 +26,7 @@ def train_one_epoch(model, optimizer, data_loader, device, epoch, print_freq, sc
 
     for images, targets in metric_logger.log_every(data_loader, print_freq, header):
         images = list(image.to(device) for image in images)
-        targets = [{k: v.to(device) for k, v in t.items()} for t in targets]
+        targets = [{k: v.to(device) if isinstance(v, torch.Tensor) else v for k, v in t.items()} for t in targets]
```
Comment (pmeier) [inline references/detection/engine.py:29]: This is for the image ID, right?

File: references/detection/presets.py [hunk 8]
```diff
@@ -1,73 +1,107 @@
...
-                    T.FixedSizeCrop(size=(1024, 1024), fill=mean),
-                    T.RandomHorizontalFlip(p=hflip_prob),
-                    T.PILToTensor(),
-                    T.ConvertImageDtype(torch.float),
-                ]
-            )
+            transforms += [
+                T.ScaleJitter(target_size=(1024, 1024), antialias=True),
+                # TODO: FixedSizeCrop below doesn't work on tensors!
+                reference_transforms.FixedSizeCrop(size=(1024, 1024), fill=mean),
```
Comment (pmeier) [inline references/detection/presets.py:46]: In v2 we have `RandomCrop` that does what `FixedSizedCrop` does minus the clamping and sanitizing bounding boxes.

File: references/detection/presets.py [hunk 9]
```diff
@@ -1,73 +1,107 @@
...
+        if backend == "pil":
+            # Note: we could just convert to pure tensors even in v2.
+            transforms += [T.ToImageTensor() if use_v2 else T.PILToTensor()]
+
+        transforms += [T.ConvertImageDtype(torch.float)]
+
+        if use_v2:
+            transforms += [
+                T.ConvertBoundingBoxFormat(datapoints.BoundingBoxFormat.XYXY),
+                T.SanitizeBoundingBox(),
```
Comment (pmeier) [inline references/detection/presets.py:79]: Do we also need `ClampBoundingBox` here?

File: references/detection/train.py [hunk 10]
```diff
@@ -177,8 +185,8 @@ def main(args):
     # Data loading code
     print("Loading data")
 
-    dataset, num_classes = get_dataset(args.dataset, "train", get_transform(True, args), args.data_path)
-    dataset_test, _ = get_dataset(args.dataset, "val", get_transform(False, args), args.data_path)
+    dataset, num_classes = get_dataset(args.dataset, "train", get_transform(True, args), args.data_path, args.use_v2)
```
Comment (pmeier) [inline references/detection/train.py:188]: Not required here, but can we maybe use keyword args here? The call is really hard to parse.

File: references/detection/transforms.py (hunk 2 above)
Comment (pmeier) [reply references/detection/transforms.py:296]: IDC. Unless there is some other opinion, let's keep it the way it is.

**Review by pmeier — APPROVED** [review]
LGTM, thanks Nicolas! I'm ok with not testing it on all configurations right now, but to make sure: you have tested it on least one and it works, correct?

**Comment by pmeier** [comment]
Test failures are real. Maybe related to `antialias`? I'll have a look.

**Comment by NicolasHug** [comment]
Thanks for the review!

Which failure are relevant? We have no tests for the `reference` folder and this PR doesn't touch any file outside of it.

File: references/detection/coco_utils.py (hunk 6 above)
Comment (NicolasHug) [reply references/detection/coco_utils.py:195]: I would if I knew what to do lol. (I'm gonna leave this out for now I think)

**Comment by pmeier** [comment]
> We have no tests for the `reference` folder and this PR doesn't touch any file outside of it.

We do. We have v2 consistency tests that checks the transforms that we have added to our package against the stuff that we have in our references:

- https://github.com/pytorch/vision/blob/08c9938f3e69b4fb2d710a038479264f3c23b133/test/test_transforms_v2_consistency.py#L1083
- https://github.com/pytorch/vision/blob/08c9938f3e69b4fb2d710a038479264f3c23b133/test/test_transforms_v2_consistency.py#L1188

I've send a fix in 72da6553e95d451ba17463436e452aaf6a6caca8.

File: references/detection/presets.py (hunk 9 above)
Comment (NicolasHug) [reply references/detection/presets.py:79]: I don't think so since we established that all transforms should clamp already (those that need to, at least)?

**Comment by NicolasHug** [comment]
> but to make sure: you have tested it on least one and it works, correct?

Yeah, I tested it on a few combinations, but I wouldn't be surprised if there's a few edge-cases I missed. We'll find out soon enough. Thanks for the review!
